Fix variable liveness for try catch in DFG

hyjorc1 opened the pull request:
#### 5e1033a7cd0361a94d7af3ca78c81873f33aa6a8
<pre>
Fix variable liveness for try catch in DFG
<a href="https://bugs.webkit.org/show_bug.cgi?id=239758">https://bugs.webkit.org/show_bug.cgi?id=239758</a>
rdar://92654142

Reviewed by Yusuke Suzuki.

There is no successor and predecessor relationship between try and catch
block in DFG, in other words node flags cannot be passed from catch block
to its `predecessors`. If a variable defined outside the try catch block
but only be used in the catch block, then our compiler would mis-analyze
the liveness of the varible w.r.t the catch block. Therefore,
`LiveCatchVariablePreservationPhase` should be to performed before backwards propagation.

* JSTests/stress/try-catch-backwards-propagation.js: Added.
(throwFunction):
(foo1):
(foo2):
(foo3):
(foo4):
(foo5):
(foo6):
(foo7):
(foo8):
(foo9):
(foo10):
(foo11):
(foo12):
(opt):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::parse):
* Source/JavaScriptCore/dfg/DFGLiveCatchVariablePreservationPhase.cpp:
(JSC::DFG::LiveCatchVariablePreservationPhase::LiveCatchVariablePreservationPhase):
(JSC::DFG::LiveCatchVariablePreservationPhase::handleBlockForTryCatch):
(JSC::DFG::performLiveCatchVariablePreservationPhase):
* Source/JavaScriptCore/dfg/DFGPlan.cpp:
(JSC::DFG::Plan::compileInThreadImpl):

Canonical link: <a href="https://commits.webkit.org/259839@main">https://commits.webkit.org/259839@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f2c8ac41013934f0fe5632c8d72338b470569b74

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106120 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15173 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/38949 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115304 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/175381 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16605 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6345 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98328 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/115000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/111876 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12629 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/95606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40178 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94502 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/27249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/81859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/95737 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8424 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/28601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/95116 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/6245 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/8918 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/5149 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/30512 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14537 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48145 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/103862 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6805 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10463 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/25739 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -9171,6 +9172,10 @@ void ByteCodeParser::parse()
     parseCodeBlock();
     linkBlocks(inlineStackEntry.m_unlinkedBlocks, inlineStackEntry.m_blockLinkingTargets);
 
+    m_graph.determineReachability();
```
Let's add comment for this performLiveCatchVariablePreservationPhase.

```
// We insert catch variable preservation here to show the all byte code uses to subsequent backward propagation phase.
```

Constellation commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -9171,6 +9172,10 @@ void ByteCodeParser::parse()
 
+    m_graph.determineReachability();
+    performLiveCatchVariablePreservationPhase(m_graph);
+    m_graph.clearReachability();
```
We should move ` m_graph.determineReachability()` and `m_graph.clearReachability()` inside `performLiveCatchVariablePreservationPhase`, and run them only when `m_graph.m_hasExceptionHandlers` is true.

Constellation reviewed (approved):
r=me with comments.

Constellation commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -9171,6 +9172,10 @@ void ByteCodeParser::parse()
+    performLiveCatchVariablePreservationPhase(m_graph);
+    m_graph.clearReachability();
+
     // We run backwards propagation now because the soundness of that phase
     // relies on seeing the graph as if it were an IR over bytecode, since
     // the spec-correctness of that phase relies on seeing all bytecode uses.
```
Literally, this is the reason why we need to insert catch preservation before backward propagation.

Constellation commented on Source/JavaScriptCore/dfg/DFGLiveCatchVariablePreservationPhase.cpp:
```diff
@@ -221,11 +223,20 @@ class LiveCatchVariablePreservationPhase : public Phase {
+    bool result = LiveCatchVariablePreservationPhase(graph).run();
+
+    if (graph.m_hasExceptionHandlers)
+        graph.clearReachability();
```
No. We should do it inside LiveCatchVariablePreservationPhase.

Constellation commented on Source/JavaScriptCore/dfg/DFGLiveCatchVariablePreservationPhase.cpp:
```diff
@@ -53,8 +53,10 @@ class LiveCatchVariablePreservationPhase : public Phase {
         InsertionSet insertionSet(m_graph);
```
See
```
        if (!m_graph.m_hasExceptionHandlers)
            return false;
```
This is why I said we should do `graph.determineReachability` only when `m_graph.m_hasExceptionHandlers` is true.
We should insert `graph.determineReachability()` here (I mean, after `m_graph.m_hasExceptionHandlers` check).

Constellation commented on Source/JavaScriptCore/dfg/DFGLiveCatchVariablePreservationPhase.cpp:
```diff
@@ -221,11 +223,20 @@ class LiveCatchVariablePreservationPhase : public Phase {
 {
-    return runPhase<LiveCatchVariablePreservationPhase>(graph);
+    if (graph.m_hasExceptionHandlers)
+        graph.determineReachability();
```
Ditto.

Constellation commented on Source/JavaScriptCore/dfg/DFGLiveCatchVariablePreservationPhase.cpp:
```diff
@@ -53,8 +53,10 @@ class LiveCatchVariablePreservationPhase : public Phase {
+                }
             }
         }
 
```
And do `graph.clearReachability()` here.

Constellation commented on Source/JavaScriptCore/dfg/DFGLiveCatchVariablePreservationPhase.cpp:
```diff
@@ -53,8 +53,10 @@ class LiveCatchVariablePreservationPhase : public Phase {
         InsertionSet insertionSet(m_graph);
         if (m_graph.m_hasExceptionHandlers) {
```
This is always true, so probably we should remove this if-condition.

hyjorc1 committed:
Fix variable liveness for try catch in DFG
https://bugs.webkit.org/show_bug.cgi?id=239758
rdar://92654142

Reviewed by Yusuke Suzuki.

There is no successor and predecessor relationship between try and catch
block in DFG, in other words node flags cannot be passed from catch block
to its `predecessors`. If a variable defined outside the try catch block
but only be used in the catch block, then our compiler would mis-analyze
the liveness of the varible w.r.t the catch block. Therefore,
`LiveCatchVariablePreservationPhase` should be to performed before backwards propagation.

* JSTests/stress/try-catch-backwards-propagation.js: Added.
(throwFunction):
(foo1):
(foo2):
(foo3):
(foo4):
(foo5):
(foo6):
(foo7):
(foo8):
(foo9):
(foo10):
(foo11):
(foo12):
(opt):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::parse):
* Source/JavaScriptCore/dfg/DFGLiveCatchVariablePreservationPhase.cpp:
(JSC::DFG::LiveCatchVariablePreservationPhase::LiveCatchVariablePreservationPhase):
(JSC::DFG::LiveCatchVariablePreservationPhase::handleBlockForTryCatch):
(JSC::DFG::performLiveCatchVariablePreservationPhase):
* Source/JavaScriptCore/dfg/DFGPlan.cpp:
(JSC::DFG::Plan::compileInThreadImpl):

Canonical link: https://commits.webkit.org/259839@main
